lib/Auth/OAuth.php

```diff
@@ -945,46 +945,32 @@ private function getOauthHeaders($includeCallback = false)
     /**
      * Normalize parameters
      *
-     * @param array  $parameters
-     * @param bool   $encode
-     * @param bool   $returnarray
-     * @param array  $normalized
-     * @param string $key
+     * @param array $parameters
+     * @param string $parentKey
      *
-     * @return array|string
+     * @return string
      */
-    private function normalizeParameters($parameters, $encode = false, $returnarray = false, $normalized = array(), $key = '')
+    private function normalizeParameters($parameters, $parentKey = '')
     {
-        // December 2016 - Fix for issue #75
-        //
-        // recursive call identified by these 2 conditions.
-        if ($returnarray && ('' != $key)) {
-            // Ref: Spec: 9.1.1 (1)
-            // If two or more parameters share the same name, they are sorted by their value
-            sort($parameters, SORT_STRING);
-        } else {
-            // Sort by key
-            ksort($parameters);
+        if (empty($parameters)) {
+            return '';
         }
 
-        foreach ($parameters as $k => $v) {
-            if (is_array($v)) {
-                $normalized = $this->normalizeParameters($v, $encode, true, $normalized, $k);
+        $normalized = array();
+        foreach ($parameters as $key => $value) {
+            if (!empty($parentKey)) {
+                // Multidimensional array; using foo=bar&foo=baz rather than foo[bar]=baz&foo[baz]=bar
+                $key = $parentKey;
+            }
+            if (is_array($value)) {
+                $normalized[] = $this->normalizeParameters($value, $key);
             } else {
-                if ($key) {
-                    //Multidimensional array; using foo=baz&foo=bar rather than foo[bar]=baz&foo[baz]=bar as this is
-                    //what the server expects when creating the signature
-                    $k = $key;
-                }
-
-                if ($encode) {
-                    $normalized[] = $this->encode($k).'='.$this->encode($v);
-                } else {
-                    $normalized[] = $k.'='.$v;
-                }
+                $normalized[] = $this->encode($key) . '=' . $this->encode($value);
```

Comment:
```
if (!empty($value)) {
    $normalized[] = $this->encode($key) . '=' . $this->encode($value);
}
```

Reply:
https://github.com/IchHabRecht/api-library/pull/1

Reply:
Hi Woeler,

My mautic installation have no lib folder (I can´t find it).

I need to solve oAuth problem in Mautic 2.14.2, so I have to modify this file. How can do this?.

Thank you.

Reply:
I posted here https://github.com/mautic/mautic/issues/5748 that I have a problem with Mautic, an oauth1a error that blocks integration with other apps.

That post brought me here, and now I can not find this file to make the modifications.

Do I have to install new libraries in my mautic installation to make oauth1 work well? Could you tell me the steps I should take?

Thank you very much.

Reply:
Once again, this has nothing to do with Mautic core. This is an independent library that allows other software to talk to the Mautic api.

You must use both Mautic version 2.14.2 and on your third party side the api library version 2.14.2 or higher.

Reply:
Thanks for your patience Woeler.

I did not want to say that I have to change anything in Mautic core, I just want to solve the problem I have with Mautic integration via API.

I understand you, but I can not find the steps to install that library. Could you help me?. I'm not a developer, but I can do the installation if you guide me a bit.

Thanks again.

Reply:
@jmarcosdleon I did not notice what integration are you talking about. That's kind of important. Does it even use this library? Contact authors of that integration and ask them for help.

Reply:
Hello scopecz,

I think this is not the place to post my mautic problem, but it has to do with the performance of oauth1 in Mautic.

I begin again to expose the problem I had with Mautic.

I was working perfectly with the Mautic API (oauth1). I had made an integration with a plugin called "ninja popus" on my wordpress website and everything was working correctly.

From the moment I updated Mautic to version 2.14.2, the integration stopped working and gave me this error in Mautic's log:

 mautic.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\HttpException: "signature_invalid" at /home2/www/email/vendor/willdurand/oauth-server-bundle/Service/OAuthServerService.php line 174 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\HttpException(code: 0): signature_invalid at /home2/www/email/vendor/willdurand/oauth-server-bundle/Service/OAuthServerService.php:174)"} []

After updating Mautic, I did not update anything else. I did not install any api library or anything like that. 

If before I worked correctly and after 2.14.2 I could not, it is that Mautic has changed. Do not you think it would be great to create an automatic update in Mautic to solve API problem and install the new libraries it need ?.

Could you help me with this? Thank you.

Reply:
I think you don't get what this library is for. It's not part of Mautic nor it should be. This library may be part of the integration. So the place what must be updated is the integration. Not Mautic. 

> I had made an integration with a plugin called "ninja popus"

Where is the source code of that integration? Can't google it. Did you create the integration yourself?

Reply:
Hi @jmarcosdleon,

Thank you for that detailed explanation of your use case. Let me try to explain why the Mautic system needed to change its code and broke your integration. While working woth oauth1 and the Mautic public API we had noticed, that not the whole API functions can be used (in combination with an oauth1 authentication process). As (of course) users rely on a working API, we needed to change a) the Mautic authentication handling and b) the Mautic API Library to be able to pass and validate all necessary API calls to a Mautic installation.

In your case, the integration ninja popus needs to be changed (updated) as well. As I don't know the plugin, I can't help with an out-of-the box solution, but I'm pretty sure you'll find some lib directory in that plugin. This lib-directory needs to be updated to the latest Mautic API Library and everything will work as before.

If there is any public available code of the plugin (or you are able to offer some directory listing) we are more then happy to even provide more steps that might help you to update your Wordpress installation.

Hope this helps to make things more understandable.

Reply:
Thank you very much IchHabRecht! I now understand.

I will contact ninja popups support to they update their plugin. Thanks again! 
If I need your help ¿where can I find you? ¿here?

Reply:
Hi @jmarcosdleon,

You can either contact me here on GitHub (I will get notifications by email) or in our Mautic Slack Channel (https://mautic.slack.com).